### PR TITLE
peers-tab: cleaner presentation - more functionality

### PR DESCRIPTION
**Pull request by RandyMcMillan**
Initial Presentation:

Screen Shot 2020-11-24 at 6 27 28 PM

peer table populated:

Screen Shot 2020-11-24 at 6 27 36 PM

Detail View of selected peer:

Screen Shot 2020-11-24 at 6 27 50 PM

Window enlarged:
Peertable and detail view:
Screen Shot 2020-11-24 at 6 30 57 PM

Additional Notes:

The widgets were renamed to offer a "self documenting" naming convention during the development.

Screen Shot 2020-11-24 at 6 39 18 PM

### CODE REVIEW

File: src/qt/forms/debugwindow.ui [hunk 1]
```diff
@@ -36,7 +36,7 @@
    <item>
     <widget class="QTabWidget" name="tabWidget">
      <property name="currentIndex">
-      <number>0</number>
+      <number>3</number>
```
Comment (luke-jr) [inline src/qt/forms/debugwindow.ui:39]: Leave this alone...

File: src/qt/forms/debugwindow.ui [hunk 2]
```diff
@@ -1141,65 +1198,121 @@
...
+                </widget>
+               </item>
+               <item row="6" column="1">
+                <widget class="QLabel" name="peerDirection">
                  <property name="cursor">
                   <cursorShape>IBeamCursor</cursorShape>
                  </property>
+                 <property name="accessibleName">
+                  <string/>
+                 </property>
```
Comment (luke-jr) [inline src/qt/forms/debugwindow.ui:1241]: blank

File: src/qt/forms/debugwindow.ui [hunk 3]
```diff
@@ -1233,42 +1339,95 @@
...
+                </widget>
+               </item>
+               <item row="5" column="1">
+                <widget class="QLabel" name="peerPermissions">
                  <property name="cursor">
                   <cursorShape>IBeamCursor</cursorShape>
                  </property>
+                 <property name="accessibleName">
+                  <string/>
+                 </property>
```
Comment (luke-jr) [inline src/qt/forms/debugwindow.ui:1392]: blank

File: src/qt/guiutil.cpp [hunk 4]
```diff
@@ -768,23 +768,66 @@ QString formatDurationStr(int secs)
...
+    });
+}
+
 QString formatServicesStr(quint64 mask)
 {
     QStringList strList;
 
     for (const auto& flag : serviceFlagsToStr(mask)) {
+        if (flag == "NETWORK_LIMITED"){
+        strList.append(QString::fromStdString("LIMITED"));
```
Comment (luke-jr) [inline src/qt/guiutil.cpp:784]: If we're going to start renaming these, "recent-blocks" makes more sense than "LIMITED"...

File: src/qt/guiutil.cpp [hunk 5]
```diff
@@ -768,23 +768,66 @@ QString formatDurationStr(int secs)
...
+        else
+        {
         strList.append(QString::fromStdString(flag));
+        }
+        sortLocaleAware(strList);
     }
 
     if (strList.size())
-        return strList.join(" & ");
+        return strList.join(" ");
```
Comment (luke-jr) [inline src/qt/guiutil.cpp:794]: This seems less clear

File: src/qt/guiutil.cpp [hunk 6]
```diff
@@ -768,23 +768,66 @@ QString formatDurationStr(int secs)
...
+        if (flag == "UNKNOWN"){
+        strList.append(QString::fromStdString("U"));
+        }
+        sortLocaleAware(strList);
+    }
+
+    if (strList.size())
+        return strList.join("");
+    else
+        return QObject::tr(""); //silence is golden
```
Comment (luke-jr) [inline src/qt/guiutil.cpp:825]: This conditional seems pointless

File: src/qt/guiutil.cpp [hunk 7]
```diff
@@ -768,23 +768,66 @@ QString formatDurationStr(int secs)
...
+        return strList.join(" ");
     else
         return QObject::tr("None");
 }
 
+QString shortFormatServicesStr(quint64 mask)
+{
+    QStringList strList;
+
+    for (const auto& flag : serviceFlagsToStr(mask)) {
```
Comment (luke-jr) [inline src/qt/guiutil.cpp:803]: Probably better to re-implement `serviceFlagsToStr` here...

**Review by luke-jr — CHANGES_REQUESTED** [review]

**Comment by jonasschnelli** [comment]
Nice. Concept ACK.

**Review by promag — COMMENTED** [review]
Concept ACK. You should split in multiple commits - renames, add more column(s), etc. Splitting in multiple commits also allows to split the PR so part  of it can be early merged. You also have unrelated changes, like in `RPCConsole::tabShortcut`, that should be removed.

File: src/qt/forms/debugwindow.ui (hunk 1 above)
[the remark this answers is not in the document]
Comment (RandyMcMillan) [reply src/qt/forms/debugwindow.ui:39]: This is Qt Creator diff junk - I will correct this if/when ready for merge.
Thanks

File: src/qt/forms/debugwindow.ui [hunk 8]
```diff
@@ -920,9 +920,27 @@
...
+             </property>
+             <property name="statusTip">
+              <string/>
+             </property>
+             <property name="whatsThis">
+              <string/>
+             </property>
+             <property name="accessibleName">
+              <string/>
+             </property>
```
[the remark this answers is not in the document]
Comment (RandyMcMillan) [reply src/qt/forms/debugwindow.ui:934]: This is Qt Creator diff junk - I will correct this if/when ready for merge.
Thanks

File: src/qt/guiutil.cpp (hunk 5 above)
Comment (RandyMcMillan) [reply src/qt/guiutil.cpp:794]: This change allows the services string in the detail view to break.
The ampersands are extraneous (imo). 

Screen Shot 2020-12-02 at 8 51 38 PM

File: src/qt/guiutil.cpp (hunk 7 above)
Comment (RandyMcMillan) [reply src/qt/guiutil.cpp:803]: The "Bumper" will be used to add more columns to the peertablemodel.
The point of having abbreviated services flags is to conserve space so more information can be displayed in the table view. The underlying design principle is to display as much useful info as possible "at a glance" as well as allowing those columns to be sortable. Connection time and synced blocks will be added in a follow up PR. 

Screen Shot 2020-12-02 at 9 03 14 PM

File: src/qt/forms/debugwindow.ui [hunk 9]
```diff
@@ -920,9 +920,30 @@
           <layout class="QVBoxLayout" name="verticalLayout_7">
            <item>
             <widget class="QTableView" name="peerWidget">
```
Comment (RandyMcMillan) [inline src/qt/forms/debugwindow.ui:922]: Upon further investigation I see that Qt Creator validates these types of files and generates these elements when Qt expects them. If they are not present when Qt expects them they generate an innocuous log - So they aren't unnecessary code - these elements are created by Qt when it expects them and is VALID code. I recommend stop considering these auto generated elements as "unnecessary code".

Screen Shot 2020-12-03 at 3 26 46 PM

**Review by RandyMcMillan — COMMENTED** [review]
Autogenerated gui elements are necessary for VALID code.

File: src/qt/forms/debugwindow.ui (hunk 3 above)
Comment (RandyMcMillan) [reply src/qt/forms/debugwindow.ui]: https://github.com/bitcoin-core/gui/pull/135#discussion_r535574885

File: src/qt/forms/debugwindow.ui [hunk 10]
```diff
@@ -1001,45 +1025,45 @@
             <height>0</height>
            </size>
           </property>
+          <property name="toolTip">
+           <string/>
+          </property>
+          <property name="statusTip">
+           <string/>
+          </property>
```
[the remark this answers is not in the document]
Comment (RandyMcMillan) [reply src/qt/forms/debugwindow.ui]: https://github.com/bitcoin-core/gui/pull/135#discussion_r535574885

File: src/qt/forms/debugwindow.ui (hunk 8 above)
[the remark this answers is not in the document]
Comment (RandyMcMillan) [reply src/qt/forms/debugwindow.ui]: https://github.com/bitcoin-core/gui/pull/135#discussion_r535574885

File: src/qt/guiutil.cpp (hunk 6 above)
Comment (RandyMcMillan) [reply src/qt/guiutil.cpp]: @luke-jr ,

- Please actually download and run a PR before reviewing.
- Before arbitrarily declaring a segment of code pointless - download the code and make the change and RUN the change.
- Feel free to post a link to what you think the code should be.

# You've done this before...

- If you aren't actually downloading and running the code - please don't review my PRs.
- I would never intentionally waste someone's time and send them down a rabbit hole.
- I don't appreciate my time being wasted.

**Comment by hebasto** [comment]
@promag 
> Concept ACK. You should split in multiple commits - renames, add more column(s), etc. Splitting in multiple commits also allows to split the PR so part of it can be early merged. You also have unrelated changes, like in `RPCConsole::tabShortcut`, that should be removed.

I'm completely agree.

@RandyMcMillan 
For example, you could split out the `I/O` column adding into a separated PR (https://github.com/bitcoin-core/gui/pull/162#issuecomment-750879842, https://github.com/bitcoin-core/gui/pull/162#pullrequestreview-558799912).

It would help make progress :)

**Comment by jonatack** [comment]
I agree with @promag and @hebasto. Smaller, more-focused commits and PRs would be a good idea here.

**Pull request closed without merging**